### ISSUE TITLE
Request - Rate limiting for failed basic auth

### ISSUE THREAD

**Issue by jeslinmx** [xuqingfeng/caddy-rate-limit#32]
There has been some interest in implementing rate limiting within the basicauth middleware to block repeated failed logins (https://github.com/mholt/caddy/issues/1318#issuecomment-269921212), but interest has shifted over to other authentication plugins such as multipass and loginsrv. 

However, there is still use for basicauth (e.g. protecting WebDAV, which is my use case), and protecting it from brute force attacks. The problem with a simple
```
ratelimit * /dav 2 3 second
webdav /dav {
    ...
}
```
is that rate limiting will be triggered not only by incorrect credentials (403 responses) but also by legitimate requests (2xx). This is a problem as one might not want to rate limit legitimate requests (or have a much higher rate limit, since certain use cases such as a client doing a sync may generate a lot of requests).

As far as I can think of, this could be done if ratelimit could be configured with a `count_if`, which determines which requests count to the request quota of an IP address, and a `block_if`, which requests will be denied for an IP address which has already exceeded its request quota and been blocked. Using , for instance, the syntax of `if` statements from the [rewrite](https://caddyserver.com/docs/rewrite) middleware:
```
ratelimit * /dav 2 3 second {
    count_if {status} is 403
webdav /dav {
    ...
}
```
Is it possible to look into this feature for caddy-rate-limit?

**Comment by xuqingfeng** [comment]
This seems a bit harder than I thought:

caddy-rate-limit needs to know the upstream status if it wants to apply limit based on the status code.

But when the request is supposed to be blocked, this plugin shouldn't pass the request to the upstream.(in this case, caddy-rate-limit won't know the upstream's status.)

**Comment by jeslinmx** [comment]
I have a method which I think will be workable; consider this example for a resource limited to 2 (403) requests:

1. user agent makes request with incorrect credentials -> plugin does not block, passes upstream -> upstream returns 403 -> plugin takes note of first 403.
2. user agent makes request with incorrect credentials -> plugin does not block, passes upstream -> upstream returns 403 -> plugin takes note of second 403, and blocks future requests.
3. user agent makes request with incorrect credentials -> plugin blocks, without passing upstream.

In other words, the condition to block a particular client only applies to before it has been blocked. After the client is blocked, all requests from that client are blocked, regardless of the condition supplied.

I'm not sure if this is a bit confusing, let me know if i need to clarify further.

**Comment by xuqingfeng** [comment]
@jeslinmx That's a good idea. 👍 

I managed to implement it in this [PR](https://github.com/xuqingfeng/caddy-rate-limit/pull/35) and merged it to master.

It would be helpful if you can try it out and give me some feedback.

(You can build caddy from source with this plugin enabled or you can use `xuqingfeng/caddy-rate-limit:latest` docker image)

**Comment by jeslinmx** [comment]
Yup, it works fantastic! Thank you so much for this!

**Comment by xuqingfeng** [comment]
FYI: this feature has been pushed to caddy website(https://caddyserver.com/docs/http.ratelimit) a few days ago. You should now be able to use the official binary directly.
